File: Pidgin.Bench.FParsec/JsonParser.fs
module Pidgin.Bench.FParsec.JsonParser

open FParsec
open Pidgin.Examples.Json
open System.Collections.Immutable

let upcastJson (x : IJson) = x

let lbrace : Parser<char, unit> = pchar '{'
let rbrace : Parser<char, unit> = pchar '}'
let lbracket : Parser<char, unit> = pchar '['
let rbracket : Parser<char, unit> = pchar ']'
let quote : Parser<char, unit> = pchar '"'
let colon : Parser<char, unit> = pchar ':'
let colonWhitespace : Parser<char, unit> = between spaces spaces colon
let comma : Parser<char, unit> = pchar ','

let string : Parser<string, unit> = between quote quote (manySatisfy (fun c -> c <> '"'))
let jsonString = string |>> JsonString |>> upcastJson

let json, jsonImpl = createParserForwardedToRef()
let jsonArray = between lbracket rbracket (sepBy (between spaces spaces json) comma) |>> ImmutableArray.CreateRange |>> JsonArray |>> upcastJson
let jsonMember = string .>> colonWhitespace .>>. json
let jsonObject = between lbrace rbrace (sepBy (between spaces spaces jsonMember) comma) |>> dict |>> ImmutableDictionary.CreateRange |>> JsonObject |>> upcastJson
do jsonImpl := jsonString <|> jsonArray <|> jsonObject

let parse s = match run json s with
                  | Success(result, _, _) -> result
                  | Failure(msg, _, _) -> failwith msg

File: Pidgin.Examples/Json/JsonParser.cs
using System.Collections.Generic;
using System.Collections.Immutable;

using static Pidgin.Parser;
using static Pidgin.Parser<char>;

namespace Pidgin.Examples.Json
{
    public static class JsonParser
    {
        private static readonly Parser<char, char> _lBrace = Char('{');
        private static readonly Parser<char, char> _rBrace = Char('}');
        private static readonly Parser<char, char> _lBracket = Char('[');
        private static readonly Parser<char, char> _rBracket = Char(']');
        private static readonly Parser<char, char> _quote = Char('"');
        private static readonly Parser<char, char> _colon = Char(':');
        private static readonly Parser<char, char> _colonWhitespace =
            _colon.Between(SkipWhitespaces);
        private static readonly Parser<char, char> _comma = Char(',');

        private static readonly Parser<char, string> _string =
            Token(c => c != '"')
                .ManyString()
                .Between(_quote);
        private static readonly Parser<char, Json> _jsonString =
            _string.Select<Json>(s => new JsonString(s));

        private static readonly Parser<char, Json> _json =
            _jsonString.Or(Rec(() => _jsonArray!)).Or(Rec(() => _jsonObject!));

        private static readonly Parser<char, Json> _jsonArray =
            _json.Between(SkipWhitespaces)
                .Separated(_comma)
                .Between(_lBracket, _rBracket)
                .Select<Json>(els => new JsonArray(els.ToImmutableArray()));

        private static readonly Parser<char, KeyValuePair<string, Json>> _jsonMember =
            _string
                .Before(_colonWhitespace)
                .Then(_json, (name, val) => new KeyValuePair<string, Json>(name, val));

        private static readonly Parser<char, Json> _jsonObject =
            _jsonMember.Between(SkipWhitespaces)
                .Separated(_comma)
                .Between(_lBrace, _rBrace)
                .Select<Json>(kvps => new JsonObject(kvps.ToImmutableDictionary()));

        public static Result<char, Json> Parse(string input) => _json.Parse(input);
    }
}
